### PR TITLE
Adapt 0.7 start menu

### PR DESCRIPTION
**Pull request by def-**
screenshot-20200904@075411
Supports hotkeys to enter them, escape to go back. Start menu can be disabled in settings. Play goes to the last tab you had open between Internet, LAN, DDNet, KoG.
Still missing:
- News picture
- More DDNet-looking Play picture
- Maybe implement the server button, but seems like too much work for little benefit

### CODE REVIEW

**Comment by def-** [comment]
code-wise ready for review. Maybe someone else wants to implement the server button. It should pipe out stdout, probably using select on Posix and some equivalent on Windows and display it inline.

**Comment by BannZay** [comment]
I would add learn button to new main menu and remove from the "old" menu setting/demos/editor/news/learn tabs making them stand alone pages with return to main menu button. 

The return button just awesome and allows you to navigate extremely fast. Exit from editor by return key is the best thing ever <3

Can we maybe just execute "start DDNet-Server.exe" if none will implement it "right" way. Will improve later. I like the Server button here because this might force user to learn stuff about ddnet server

**Comment by BannZay** [comment]
Nice Deen, this is very as is already. I love it

**Comment by def-** [comment]
> I would add learn button to new main menu and remove from the "old" menu setting/demos/editor/news/learn tabs making them stand alone pages with return to main menu button.

Done

> Can we maybe just execute "start DDNet-Server.exe" if none will implement it "right" way. Will improve later. I like the Server button here because this might force user to learn stuff about ddnet server

Done, but F7 doesn't work properly on Windows and I don't see a standard way of opening a terminal window in Linux.

File: src/game/client/components/menus_start.cpp [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+		if(m_ServerProcess.Process)
+		{
+			kill_process(m_ServerProcess.Process);
+			m_ServerProcess.Process = 0;
+		}
+		else
+		{
+			Storage()->RemoveBinaryFile("autoexec_server.log");
+			char aBuf[MAX_PATH_LENGTH];
+			m_ServerProcess.Process = shell_execute(Storage()->GetBinaryPath(PLAT_SERVER_EXEC, aBuf, sizeof(aBuf)));
```
Comment (BannZay) [inline src/game/client/components/menus_start.cpp:63]: yea, it wont work on windows if working directory is not where DDNet-Server.exe located as storage contains wrong data IMO

I failed to understand the code of FindDatadir which assumes that binary dir are always in the same directory where data directory located. 

At least CMAKE builds following structure for windows:
build\Debug\DDnet.exe 
build\data\mapres

I see no reason why we just wont get the full directory path to executive assembly for m_aBinarydir? (GetModuleFileNameA for windows)

File: src/game/client/components/menus.cpp [hunk 2]
```diff
@@ -684,99 +715,103 @@ int CMenus::RenderMenubar(CUIRect r)
...
-		{
-			NewPage = PAGE_NEWS;
-			m_DoubleClickIndex = -1;
-		}
-		Box.VSplitLeft(60.0f, &Button, &Box);
-		static int s_LearnButton=0;
-		if(DoButton_MenuTab(&s_LearnButton, Localize("Learn"), false, &Button, CUI::CORNER_TR))
+		Box.VSplitLeft(100.0f, &Button, &Box);
+		static int s_StartButton=0;
+		if(DoButton_MenuTab(&s_StartButton, Localize("Back"), false, &Button, CUI::CORNER_T))
```
Comment (BannZay) [inline src/game/client/components/menus.cpp:728]: "Menu" instead of "Back" looks better to me

**Comment by edg-l** [comment]
This is also a chance to add a discord button, since now there is space for it i think.

File: src/game/client/components/menus.cpp (hunk 2 above)
Comment (def-) [reply src/game/client/components/menus.cpp:728]: Done.

**Comment by def-** [comment]
> This is also a chance to add a discord button, since now there is space for it i think.

I don't think that belongs in the client.

**Comment by Ekidoxx** [comment]
Three buttons in bottom left corner - for website, wiki and discord could be nice addition actually.

**Comment by BannZay** [comment]
Should not we move learn and news to bottom left as well? They does not have icons & logically not really part of actual gameplay

File: src/base/system.c [hunk 3]
```diff
@@ -3242,10 +3254,17 @@ int pid(void)
...
 }
 
-void shell_execute(const char *file)
+PROCESS shell_execute(const char *file)
 {
 #if defined(CONF_FAMILY_WINDOWS)
-	ShellExecute(NULL, NULL, file, NULL, NULL, SW_SHOWDEFAULT);
+	SHELLEXECUTEINFOA info;
+	mem_zero(&info, sizeof(SHELLEXECUTEINFOA));
+	info.lpVerb = "open";
```
[the remark this answers is not in the document]
Comment (BannZay) [reply src/base/system.c:3262]: @def- Does not work without cbsize (this is required param)

File: src/game/client/components/menus.cpp [hunk 4]
```diff
@@ -2011,6 +2065,27 @@ extern "C" void font_debug_render();
 
 void CMenus::OnRender()
 {
+	if(m_ServerProcess.Process != 0)
+	{
+		// TODO: Maybe call this less often, or probably even cleaner, use econ
+		if(!m_ServerProcess.Initialized)
+		{
+			char aBuf[MAX_PATH_LENGTH];
+			IOHANDLE File = Storage()->OpenFile(Storage()->GetBinaryPath("autoexec_server.log", aBuf, sizeof(aBuf)), IOFLAG_READ, IStorage::TYPE_ALL);
```
Comment (heinrich5991) [inline src/game/client/components/menus.cpp:2074]: Does this work on Windows? It usually has problems with two processes accessing the same file.

**Comment by heinrich5991** [comment]
How does the econ console work? I can't find it in the code.

Looks nice. Should we get this in soon or wait a little more?

File: src/game/client/components/menus.cpp (hunk 4 above)
Comment (def-) [reply src/game/client/components/menus.cpp:2074]: I will get rid of this. I noticed that Windows didn't work, probably some io_seek magic would fix it, but econ console seems more stable.

**Comment by def-** [comment]
> How does the econ console work? I can't find it in the code.

It doesn't work yet. I still have to implement the EconClient.

**Comment by heinrich5991** [comment]
I'm fine with waiting a little.

**Comment by def-** [comment]
screenshot-20200907@120453

File: src/base/system.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (def-) [reply src/base/system.c:3222]: Done.

File: src/game/client/components/menus_start.cpp (hunk 1 above)
Comment (def-) [reply src/game/client/components/menus_start.cpp:107]: But we never release builds like that, right? The autoupdater also depends on this location of DDNet-Server.exe, so I think we can keep it as is. The alternative would be to offer something like FindDatadir but FindDDNetServer that searches in some paths, starting with pArgv0. For Mac OSX the path is probably also wrong, not sure how to find it there. @Learath2 any idea, should we just hardcode it?

Comment (BannZay) [reply src/game/client/components/menus_start.cpp:107]: I find it good just to store our exe's in `binary` dir. We should probably explain this idea to CMAKE

Comment (def-) [reply src/game/client/components/menus_start.cpp:107]: btw, my build doesn't create Debug directory, maybe cause I build in-directory with `cmake .`

Comment (BannZay) [reply src/game/client/components/menus_start.cpp:107]: But still, maybe it is better to query an executable location and mark its folder as binary folder. Is binary files should be a part of storage.cfg ?

**Comment by def-** [comment]
@heinrich5991 @Learath2 any suggestion how to find DDNet-Server.exe on OSX distribution and Windows cmake build?

Other than that should be done, I removed the econ for now since I got annoyed by it.

**Comment by BannZay** [comment]
> any suggestion how to find DDNet-Server.exe

GetModuleFileNameA can be used for windows. As client and server builds into the same directory

**Comment by def-** [comment]
That does *not* work for Mac OSX. If we use GetModuleFileNameA we could also just use argv[0] which is available on more platforms, couldn't we?

**Comment by def-** [comment]
screenshot-20200907@204748

**Comment by BannZay** [comment]
if you click on News or demos and then on Settings tab then you will be navigated to Play menu
I guess we should remove settings and Editor from tabs bar

**Comment by def-** [comment]
I actually like to still have the small buttons there for settings and quit. How Teeworlds solved it with remembering the previous page when you go back from settings might be a bit cleaner, but I'm ok with our current solution.

**Comment by Ekidoxx** [comment]
Little bug with the tabs

1. Click on `Demos`
image

2. From there go to `settings`
image

3. \*Puff\* `Demos` tab changed to `Internet`, `LAN`, `Favourites`, `DDNet` & `KoG` tabs
Or otherwise said, you're moved from `Demos` to `Play`
image


Discovered thanks to @BannZay

**Comment by def-** [comment]
I don't actually mind that behaviour much. Not sure if it's worth the effort to change it. In the end after changing settings people mostly go to play :D

**Comment by BannZay** [comment]
This menus is so fresh for DDNet.

I would make this menu the only thing you can use to reach places without using bindings.

I can continue the work on tabs after this pr merge if you dont mind.

**Comment by def-** [comment]
@BannZay Sure, would be great if you can improve it a bit. I'm a bit tired of UI changes, so from my point of view this should go in.

**Comment by Ekidoxx** [comment]
What do you guys think about our own, DDNet-related images taken directly out of the game & client?

image

**Comment by BannZay** [comment]
Idk, its cool to have own and they are descent, but teeworlds ones way cooler tho

**Comment by kingston-svg** [comment]
It's cool, I like it.

**Comment by HiRavie** [comment]
I suggest an entire new layout to set us apart from 0.7, I sketched this one up. It can surely still be improved, but let me know what you think before I put more time into this. 

layout

**Review by heinrich5991 — CHANGES_REQUESTED** [review]
Bugs found while testing:
- The loading bar goes all the way to the right before starting at 0 and going to 100%.
- Run server hangs the client.
- Quit asks whether I really want to quit (as opposed to usually where it only asks it when you're ingame for more than 20 min).

**Comment by def-** [comment]
> The loading bar goes all the way to the right before starting at 0 and going to 100%.

What loading bar?

> Run server hangs the client.

Any idea how to reproduce? Works for me.

> Quit asks whether I really want to quit (as opposed to usually where it only asks it when you're ingame for more than 20 min).

Fixed.

**Comment by heinrich5991** [comment]
I get
```
[xcb] Unknown sequence number while processing queue
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
DDNet: xcb_io.c:260: poll_for_event: Assertion `!xcb_xlib_threads_sequence_lost' failed.
```

When running with strace, I don't get the hang, but still get these messages…

The loading bar I'm referring to is the initial loading bar of the DDNet client.

**Comment by def-** [comment]
> The loading bar I'm referring to is the initial loading bar of the DDNet client.

That bar is one frame for me so impossible to debug.

**Comment by Learath2** [comment]
Slow it down arbitrarily, or maybe record it?

**Comment by def-** [comment]
I figured it out already, easy enough to fix.

But no idea about the other problem with running DDNet-Server yet

File: src/base/system.c [hunk 5]
```diff
@@ -3214,6 +3234,18 @@ void shell_execute(const char *file)
 	pid = fork();
 	if(!pid)
 		execv(file, argv);
```
Comment (heinrich5991) [inline src/base/system.c:3236]: Found it, happens on file not found for the server.

This seems to happen due to incorrect error checking. When execv returns, process creation failed and the current process must be killed.

`fork()` should probably also get some error checking.

Comment (def-) [reply src/base/system.c:3241]: Thanks, done.

**Review by heinrich5991 — APPROVED** [review]
Thanks, looks good now.

bors r+

**Comment by def-** [comment]
@HiRavie maybe you can work with @BannZay if you two still want to improve the layout.

**Pull request merged**